server/src/main/java/org/elasticsearch/index/mapper/TimeSeriesModeIdFieldMapper.java

```diff
@@ -0,0 +1,140 @@
+            return termsQuery(Arrays.asList(value), context);
+        }
+
+        @Override
+        public Query existsQuery(SearchExecutionContext context) {
+            return new MatchAllDocsQuery();
+        }
+
+        @Override
+        public Query termsQuery(Collection<?> values, SearchExecutionContext context) {
+            failIfNotIndexed();
+            BytesRef[] bytesRefs = values.stream().map(v -> {
+                Object idObject = v;
+                if (idObject instanceof BytesRef) {
+                    idObject = ((BytesRef) idObject).utf8ToString();
+                }
+                return Uid.encodeId(idObject.toString());
+            }).toArray(BytesRef[]::new);
+            return new TermInSetQuery(name(), bytesRefs);
+        }
+    }
+
+    private TimeSeriesModeIdFieldMapper() {
+        super(new IdFieldType(), Lucene.KEYWORD_ANALYZER);
+    }
+
+    private static final long SEED = 0;
+
+    public void createField(DocumentParserContext context, BytesRef tsid) {
+        IndexableField[] timestampFields = context.rootDoc().getFields(DataStreamTimestampFieldMapper.DEFAULT_PATH);
+        if (timestampFields.length == 0) {
+            throw new IllegalArgumentException(
+                "data stream timestamp field [" + DataStreamTimestampFieldMapper.DEFAULT_PATH + "] is missing"
+            );
+        }
+        long timestamp = timestampFields[0].numericValue().longValue();
+
+        Hash128 hash = new Hash128();
+        MurmurHash3.hash128(tsid.bytes, tsid.offset, tsid.length, SEED, hash);
+
+        byte[] suffix = new byte[16];
+        ByteUtils.writeLongLE(hash.h1, suffix, 0);
+        ByteUtils.writeLongLE(timestamp, suffix, 8);   // TODO compare disk usage for LE and BE on timestamp
+
+        IndexRouting.ExtractFromSource indexRouting = (IndexRouting.ExtractFromSource) context.indexSettings().getIndexRouting();
+        // TODO it'd be way faster to use the fields that we've extract here rather than the source or parse the tsid
+        context.id(indexRouting.createId(context.sourceToParse().getXContentType(), context.sourceToParse().source(), suffix));
+        assert Uid.isURLBase64WithoutPadding(context.id()); // Make sure we get to use Uid's nice optimizations
+        /*
+         * Make sure that _id from extracting the tsid matches that _id
+         * from extracting the _source. This should be true for all valid
+         * documents with valid mappings. *But* some invalid mappings
+         * will not parse the field but be rejected later by the dynamic
+         * mappings machinery. So if there are any dynamic mappings
+         * at all we just skip the assertion because we can't be sure
+         * it always must pass.
+         */
+        assert context.getDynamicMappers().isEmpty() == false
+            || context.getDynamicRuntimeFields().isEmpty() == false
+            || context.id().equals(indexRouting.createId(TimeSeriesIdFieldMapper.decodeTsid(tsid), suffix));
```

Comment:
I'm not really confident in this assert. The *vast* majority of the time we can just check that that _id is equal. But there are weird invalid config that'll be rejected later that fail that assertion.

I'm not sure if it improves test coverage to change from the _source generated routing to the decoded _tsid. I added a bunch of unit tests that assert that they are the same and a few yaml tests. But this assertion really is the strongest assurance we have.

It would make it a bit faster to switch them though. But, like you said, I'm going to try and replace this in a follow up anyway.